python/packages/zarathustra/skills/subgraph_query_abci/behaviours.py

```diff
@@ -64,40 +68,73 @@ class CheckSubgraphsHealthBehaviour(SubgraphQueryBaseBehaviour):
+            ledger_block_number = 20_000_000  # TODO
+            # ledger_block_number = yield from self.get_latest_block_from_ledger()
+            self.context.logger.info(f"Latest ledger block: {ledger_block_number}")
+            subgraph_block_number = self.get_latest_block_from_subgraph()
+            self.context.logger.info(f"Latest subgraph block: {subgraph_block_number}")
+
+            # as subgraph is often slightly behind, we use a tolerance threshold
+            tolerance = 10
+            subgraph_health = ledger_block_number - subgraph_block_number < tolerance
+            payload = CheckSubgraphsHealthPayload(sender=sender, content=subgraph_health)
 
         with self.context.benchmark_tool.measure(self.behaviour_id).consensus():
             yield from self.send_a2a_transaction(payload)
             yield from self.wait_until_round_end()
 
-        self.set_done()
+    def get_latest_block_from_subgraph(self) -> int:
+        api_key = self.params.config["subgraph_api_key"]
+        subgraph_config = json.loads(self.synchronized_data.most_voted_subgraph_config)
+        subgraph_url = subgraph_config["subgraph_url"]
+        query = "{_meta {block {number}}}"
+        url = subgraph_url.format(api_key=api_key)
+        headers = {"Content-Type": "application/json"}
+        response = requests.post(url, json={"query": query}, headers=headers)
+        subgraph_block_number = response.json()["data"]["_meta"]["block"]["number"]
+        return subgraph_block_number
+
+    def get_latest_block_from_ledger(self):
+        ledger_api_response = yield from self.get_ledger_api_response(
+            performative=LedgerApiMessage.Performative.GET_STATE,
+            ledger_callable="get_block",
+            block_identifier="latest",
+        )
+        self.context.logger.info(f"Ledger API response: {ledger_api_response}")
+        return ledger_api_response
 
 
 class CollectSubgraphsDataBehaviour(SubgraphQueryBaseBehaviour):
     """CollectSubgraphsDataBehaviour"""
 
     matching_round: Type[AbstractRound] = CollectSubgraphsDataRound
 
-    # TODO: implement logic required to set payload content for synchronization
     def async_act(self) -> Generator:
         """Do the act, supporting asynchronous execution."""
 
         with self.context.benchmark_tool.measure(self.behaviour_id).local():
             sender = self.context.agent_address
-            payload = CollectSubgraphsDataPayload(sender=sender, content="dummy_content")
+            api_key = self.params.config["subgraph_api_key"]
+            subgraph_config = json.loads(self.synchronized_data.most_voted_subgraph_config)
+            subgraph_url = subgraph_config["subgraph_url"]
+            query = subgraph_config["subgraph_query"]
+            url = subgraph_url.format(api_key=api_key)
+            headers = {"Content-Type": "application/json"}
+            response = requests.post(url, json={"query": query}, headers=headers)
```

Comment:
currently there are several calls being made with the `requests` library, which we want to replace with message passing to the `http_client` connection

Reply:
Yess we need to convert them to the http client library.

Reply:
(mine oc haha

Reply:
not entirely sure which of the two it uses, but it works